src/fvm_multicell.hpp

```diff
@@ -117,6 +139,19 @@ class fvm_multicell {
         return min_remaining_steps_==0;
     }
 
+    // Access to sample data post-integration.
+    decltype(memory::make_const_view(std::declval<host_view>())) sample_value() const {
+        EXPECTS(!sample_events_ || sample_events_->empty());
+        host_sample_value_ = memory::on_host(sample_value_);
```

Comment:
The primary reason for the current approach lies in the differing requirements (copy, no copy) for the gpu and multicore backend implementations respectively.

That said, we shouldn't have any synchronization issues per se:
1. The `host_sample_value_` object is private, and can only be accessed via the `sample_value()` method, which assigns to it. As access to the `fvm_multicell` object is always single-threaded, this in turn means that any public access will return a faithful representation of the underlying `sample_value_` host- or device-side array.
2. If the methods are called while an integration is in progress, the values in the array may be unspecified (there can be pending writes to values yet to be performed in a later integration step), but it will not actually cause an undefined behaviour etc.
3. The comment and the assertion describe the correct pre-condition for calling.

Having an actual cached vector, like for `time_`, would mean we perform a useless copy on the multicore implementation, and makes the implementation a bit more complicated (we would need to add invalidation points, tests etc. like we do for `time_` access.)

The integration loop, comprising the call to `setup_integration` and the subsequent calls to `step_integration` can now be wrapped in a simple interface on the lowered cell, e.g. called `run_integration` or `integrate`. This would simplify the preconditions on the sample access functions, and also allow us to remove the `min_time`, `max_time` etc. methods. There are some tests though that use the `step_integration` interface.

How about I make a separate task for this `run_integration` method and associated tidying up of the lowered cell API?